service/topology_coordinator.cc

```diff
@@ -67,6 +68,121 @@ future<inet_address> wait_for_ip(raft::server_id id, const raft_address_map& am,
+    // we can remove them by their index
+    std::sort(host_load.begin(), host_load.begin() + to_remove, cmp_by_idx);
+
+    for (size_t i = 0; i < to_remove; i++) {
+        replicas.erase(replicas.begin() + host_load[i].first);
+    }
+}
+
+static void upsize_tablet_replicas_to_rf(locator::tablet_replica_set& replicas, size_t current_rf,
+    size_t new_rf, locator::load_sketch& load, const locator::token_metadata_ptr& tm, const sstring& dc)
+{
+    for (const auto& endpoint : tm->get_topology().get_datacenter_endpoints().at(dc)) {
+        locator::host_id ep = tm->get_topology().find_node(endpoint)->host_id();
+        replicas.emplace_back(locator::tablet_replica{ep, load.next_shard(ep)});
+        current_rf++;
+        if (current_rf == new_rf)
+            break;
+    }
+}
+
+future<locator::tablet_map>
+reallocate_tablets_for_new_rf(schema_ptr s, locator::token_metadata_ptr tm,
+    std::unordered_map<sstring, size_t> new_dc_rep_factor) {
+    // FIXME: Don't use tokens to distribute nodes.
+    // The following reuses the existing token-based algorithm used by NetworkTopologyStrategy.
+    assert(!tm->sorted_tokens().empty());
+
+    auto table_id = s->id();
+    locator::tablet_map tablets = tm->tablets().get_tablet_map(table_id);
+
+    locator::load_sketch load(tm);
+    co_await load.populate();
+
+    auto host_id_to_dc = [&tm](const locator::host_id& ep) -> std::optional<sstring> {
+        auto node = tm->get_topology().find_node(ep);
+        if (node == nullptr) {
+            return std::nullopt;
+        }
+        return node->dc_rack().dc;
+    };
+
+    for (auto tb : tablets.tablet_ids()) {
+        const locator::tablet_info& ti = tablets.get_tablet_info(tb);
+        std::unordered_map<sstring, size_t> dc_rep_factor;
+        locator::tablet_replica_set replicas = ti.replicas;
+
+        for (auto&& replica : replicas) {
+            dc_rep_factor[*host_id_to_dc(replica.host)]++;
+        }
+        for (const auto& [dc, rf] : new_dc_rep_factor) {
+            // ensure that we have keys for all DCs
+            dc_rep_factor[dc];
+        }
+
+        for (auto& [dc, rf] : dc_rep_factor) {
+            co_await coroutine::maybe_yield();
+
+            auto new_rf = new_dc_rep_factor[dc];
+
+            // skip DCs that are not in the new RF per DC map
+            if (!new_dc_rep_factor.contains(dc))
+                continue;
+
+            // skip DCs that have the same RF
+            if (rf == new_rf)
+                continue;
+
+            if (rf > new_rf) {
+                // remove replicas
+                downsize_tablet_replicas_to_rf(replicas, rf, new_rf, load, dc, host_id_to_dc);
+            } else {
+                // add replicas
+                upsize_tablet_replicas_to_rf(replicas, rf, new_rf, load, tm, dc);
+            }
+        }
+
+        tablets.set_tablet(tb, locator::tablet_info{std::move(replicas)});
+    }
+
+    locator::tablet_logger.debug("Reallocated tablets for {}.{} ({}): {}",
```

Comment:
This log can be removed altogether for 2 reasons: first, we're not reallocating, just adding/removing, 2nd - actually we're just creating some new memory objects and the real action will be executed elsewhere.